src/main/java/com/epam/eco/commons/kafka/helpers/CachedTopicRecordFetcher.java

```diff
@@ -0,0 +1,339 @@
+/*******************************************************************************
+ *  Copyright 2023 EPAM Systems
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ *  use this file except in compliance with the License.  You may obtain a copy
+ *  of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ *  License for the specific language governing permissions and limitations under
+ *  the License.
+ *******************************************************************************/
+package com.epam.eco.commons.kafka.helpers;
+
+import java.time.OffsetDateTime;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Predicate;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.common.TopicPartition;
+
+import com.epam.eco.commons.kafka.KafkaUtils;
+import com.epam.eco.commons.kafka.OffsetRange;
+import static java.time.OffsetDateTime.now;
+import static java.util.Objects.isNull;
+
+/**
+ * @author Mikhail_Vershkov
+ */
+public class CachedTopicRecordFetcher<K, V> extends BiDirectionalTopicRecordFetcher<K,V> implements RecordBiDirectionalFetcher<K,V> {
```

Comment:
Looks dangerous for usage from resource consumption point of view.
Maybe move this into incubator/poc package or get rid of it?